File: rust/crates/eckit/src/stream.rs
//! Stream — tagged binary serialization over TCP or memory.
//!
//! Wraps `eckit::Stream` (`TCPStream`, `MemoryStream`, `ResizableMemoryStream`).
//! Used by the Hermes protocol for `MarsRequest` serialization.

use crate::error::Result;

/// Tagged binary stream — primitive read/write operations.
///
/// Implemented by [`TcpStream`] and [`MemoryStream`].
/// Hermes protocol code uses `&mut dyn Stream` for testability.
pub trait Stream {
    fn write_u8(&mut self, v: u8) -> Result<()>;
    fn write_bool(&mut self, v: bool) -> Result<()>;
    fn write_i32(&mut self, v: i32) -> Result<()>;
    fn write_i64(&mut self, v: i64) -> Result<()>;
    fn write_u64(&mut self, v: u64) -> Result<()>;
    fn write_f64(&mut self, v: f64) -> Result<()>;
    fn write_string(&mut self, v: &str) -> Result<()>;
    fn write_blob(&mut self, data: &[u8]) -> Result<()>;

    fn read_u8(&mut self) -> Result<u8>;
    fn read_bool(&mut self) -> Result<bool>;
    fn read_i32(&mut self) -> Result<i32>;
    fn read_i64(&mut self) -> Result<i64>;
    fn read_u64(&mut self) -> Result<u64>;
    fn read_f64(&mut self) -> Result<f64>;
    fn read_string(&mut self) -> Result<String>;

    fn bytes_written(&self) -> i64;

    /// Access the underlying C++ stream wrapper (for cross-crate FFI).
    fn as_sys_mut(&mut self) -> std::pin::Pin<&mut eckit_sys::StreamWrapper>;
}

/// Serialize a value to a stream.
pub trait StreamWrite {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()>;
}

/// Deserialize a value from a stream.
pub trait StreamRead: Sized {
    fn read_from(stream: &mut dyn Stream) -> Result<Self>;
}

// ==================== Macro for Stream impl delegation ====================

/// Implements [`Stream`] by delegating to the inner `eckit_sys::StreamWrapper`.
macro_rules! impl_stream {
    ($ty:ty) => {
        impl Stream for $ty {
            fn write_u8(&mut self, v: u8) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_char(v)
                    .map_err(eckit_sys::Error::from)
            }
            fn write_bool(&mut self, v: bool) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_bool(v)
                    .map_err(eckit_sys::Error::from)
            }
            fn write_i32(&mut self, v: i32) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_int(v)
                    .map_err(eckit_sys::Error::from)
            }
            fn write_i64(&mut self, v: i64) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_long(v)
                    .map_err(eckit_sys::Error::from)
            }
            fn write_u64(&mut self, v: u64) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_unsigned_long(v)
                    .map_err(eckit_sys::Error::from)
            }
            fn write_f64(&mut self, v: f64) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_double(v)
                    .map_err(eckit_sys::Error::from)
            }
            fn write_string(&mut self, v: &str) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_string(v)
                    .map_err(eckit_sys::Error::from)
            }
            fn write_blob(&mut self, data: &[u8]) -> Result<()> {
                self.inner
                    .pin_mut()
                    .write_blob(data)
                    .map_err(eckit_sys::Error::from)
            }
            fn read_u8(&mut self) -> Result<u8> {
                self.inner
                    .pin_mut()
                    .read_char()
                    .map_err(eckit_sys::Error::from)
            }
            fn read_bool(&mut self) -> Result<bool> {
                self.inner
                    .pin_mut()
                    .read_bool()
                    .map_err(eckit_sys::Error::from)
            }
            fn read_i32(&mut self) -> Result<i32> {
                self.inner
                    .pin_mut()
                    .read_int()
                    .map_err(eckit_sys::Error::from)
            }
            fn read_i64(&mut self) -> Result<i64> {
                self.inner
                    .pin_mut()
                    .read_long()
                    .map_err(eckit_sys::Error::from)
            }
            fn read_u64(&mut self) -> Result<u64> {
                self.inner
                    .pin_mut()
                    .read_unsigned_long()
                    .map_err(eckit_sys::Error::from)
            }
            fn read_f64(&mut self) -> Result<f64> {
                self.inner
                    .pin_mut()
                    .read_double()
                    .map_err(eckit_sys::Error::from)
            }
            fn read_string(&mut self) -> Result<String> {
                self.inner
                    .pin_mut()
                    .read_string()
                    .map_err(eckit_sys::Error::from)
            }
            fn bytes_written(&self) -> i64 {
                self.inner.bytes_written()
            }
            fn as_sys_mut(&mut self) -> std::pin::Pin<&mut eckit_sys::StreamWrapper> {
                self.inner.pin_mut()
            }
        }
    };
}

// ==================== TcpStream ====================

/// TCP stream — connects to a Hermes server.
pub struct TcpStream {
    inner: eckit_sys::UniquePtr<eckit_sys::StreamWrapper>,
}

#[allow(clippy::non_send_fields_in_send_ty)]
unsafe impl Send for TcpStream {}

impl TcpStream {
    /// Connect to a TCP host:port.
    pub fn connect(host: &str, port: i32) -> Result<Self> {
        let inner = eckit_sys::stream_connect(host, port).map_err(eckit_sys::Error::from)?;
        Ok(Self { inner })
    }

    /// Read raw bytes from the underlying socket (for data transfer after protocol handshake).
    pub fn read_bytes(&mut self, buf: &mut [u8]) -> Result<i64> {
        self.inner
            .pin_mut()
            .read_bytes(buf)
            .map_err(eckit_sys::Error::from)
    }
}

/// Hand off the connected socket as a streaming `DataHandle`.
///
/// Use after a protocol handshake when the remainder of the connection should
/// be drained as raw bytes — the resulting `DataHandle` reads on demand from
/// the socket, with no in-memory buffering. Consumes the [`TcpStream`].
impl TryFrom<TcpStream> for crate::DataHandle {
    type Error = crate::Error;

    fn try_from(mut stream: TcpStream) -> Result<Self> {
        let inner = stream
            .inner
            .pin_mut()
            .into_data_handle()
            .map_err(eckit_sys::Error::from)?;
        Ok(Self::from_raw(inner))
    }
}

impl_stream!(TcpStream);

// ==================== MemoryStream ====================

/// Memory-backed stream — for testing and in-memory serialization.
pub struct MemoryStream {
    inner: eckit_sys::UniquePtr<eckit_sys::StreamWrapper>,
}

#[allow(clippy::non_send_fields_in_send_ty)]
unsafe impl Send for MemoryStream {}

impl MemoryStream {
    /// Create a resizable memory stream for writing.
    #[must_use]
    pub fn writer() -> Self {
        Self {
            inner: eckit_sys::stream_memory_write(),
        }
    }

    /// Create a memory stream for reading from existing data.
    #[must_use]
    pub fn reader(data: &[u8]) -> Self {
        Self {
            inner: eckit_sys::stream_memory_read(data),
        }
    }

    /// Get the written data (writer streams only).
    pub fn buffer(&mut self) -> Result<&[u8]> {
        self.inner
            .pin_mut()
            .buffer()
            .map_err(eckit_sys::Error::from)
    }
}

impl_stream!(MemoryStream);

// ==================== StreamWrite impls for primitive types ====================

impl StreamWrite for u8 {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_u8(*self)
    }
}

impl StreamWrite for bool {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_bool(*self)
    }
}

impl StreamWrite for i32 {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_i32(*self)
    }
}

impl StreamWrite for i64 {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_i64(*self)
    }
}

impl StreamWrite for u64 {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_u64(*self)
    }
}

impl StreamWrite for f64 {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_f64(*self)
    }
}

impl StreamWrite for str {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_string(self)
    }
}

impl StreamWrite for String {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_string(self)
    }
}

impl StreamWrite for [u8] {
    fn write_to(&self, stream: &mut dyn Stream) -> Result<()> {
        stream.write_blob(self)
    }
}

// ==================== StreamRead impls for primitive types ====================

impl StreamRead for u8 {
    fn read_from(stream: &mut dyn Stream) -> Result<Self> {
        stream.read_u8()
    }
}

impl StreamRead for bool {
    fn read_from(stream: &mut dyn Stream) -> Result<Self> {
        stream.read_bool()
    }
}

impl StreamRead for i32 {
    fn read_from(stream: &mut dyn Stream) -> Result<Self> {
        stream.read_i32()
    }
}

impl StreamRead for i64 {
    fn read_from(stream: &mut dyn Stream) -> Result<Self> {
        stream.read_i64()
    }
}

impl StreamRead for u64 {
    fn read_from(stream: &mut dyn Stream) -> Result<Self> {
        stream.read_u64()
    }
}

impl StreamRead for f64 {
    fn read_from(stream: &mut dyn Stream) -> Result<Self> {
        stream.read_f64()
    }
}

impl StreamRead for String {
    fn read_from(stream: &mut dyn Stream) -> Result<Self> {
        stream.read_string()
    }
}

File: rust/crates/eckit-sys/cpp/stream.cpp
// eckit Stream bridge — implementation.

#include "eckit_exceptions.h"

#include "eckit-sys/src/lib.rs.h"
#include "stream.h"

#include "eckit/exception/Exceptions.h"
#include "eckit/io/TCPSocketHandle.h"
#include "eckit/net/TCPClient.h"

namespace eckit_bridge {

// ==================== Stream (base class) ====================

void StreamWrapper::write_char(uint8_t c) {
    stream() << static_cast<char>(c);
}
void StreamWrapper::write_bool(bool v) {
    stream() << v;
}
void StreamWrapper::write_int(int32_t v) {
    stream() << v;
}
void StreamWrapper::write_long(int64_t v) {
    stream() << static_cast<long long>(v);
}
void StreamWrapper::write_unsigned_long(uint64_t v) {
    stream() << static_cast<unsigned long long>(v);
}
void StreamWrapper::write_double(double v) {
    stream() << v;
}
void StreamWrapper::write_string(rust::Str v) {
    stream() << std::string(v.data(), v.size());
}
void StreamWrapper::write_blob(rust::Slice<const uint8_t> data) {
    stream().writeBlob(data.data(), data.size());
}

uint8_t StreamWrapper::read_char() {
    char c{};
    stream() >> c;
    return static_cast<uint8_t>(c);
}
bool StreamWrapper::read_bool() {
    bool v{};
    stream() >> v;
    return v;
}
int32_t StreamWrapper::read_int() {
    int v{};
    stream() >> v;
    return v;
}
int64_t StreamWrapper::read_long() {
    long long v{};
    stream() >> v;
    return static_cast<int64_t>(v);
}
uint64_t StreamWrapper::read_unsigned_long() {
    unsigned long long v{};
    stream() >> v;
    return static_cast<uint64_t>(v);
}
double StreamWrapper::read_double() {
    double v{};
    stream() >> v;
    return v;
}
rust::String StreamWrapper::read_string() {
    std::string v;
    stream() >> v;
    return rust::String(v);
}

int64_t StreamWrapper::read_bytes(rust::Slice<uint8_t> buf) {
    throw eckit::SeriousBug("read_bytes not supported on this stream type");
}

rust::Slice<const uint8_t> StreamWrapper::buffer() {
    throw eckit::SeriousBug("buffer() not supported on this stream type");
}

int64_t StreamWrapper::bytes_written() const {
    return const_cast<StreamWrapper*>(this)->stream().bytesWritten();
}

std::unique_ptr<DataHandleWrapper> StreamWrapper::into_data_handle() {
    throw eckit::NotImplemented("StreamWrapper::into_data_handle is only supported on TCP streams", Here());
}

// ==================== TcpStreamWrapper ====================

TcpStreamWrapper::TcpStreamWrapper(const std::string& host, int port) :
    socket_(eckit::net::TCPClient().connect(host, port)), tcp_(socket_) {}

int64_t TcpStreamWrapper::read_bytes(rust::Slice<uint8_t> buf) {
    // The connection lives on `tcp_.socket()` — `socket_` was emptied by the
    // TCPSocket "copy" ctor in `tcp_(socket_)` (ownership transfer).
    return tcp_.socket().read(buf.data(), static_cast<long>(buf.size()));
}

std::unique_ptr<DataHandleWrapper> TcpStreamWrapper::into_data_handle() {
    // Steal the live connection from `tcp_` and wrap it as an owning
    // `eckit::TCPSocketHandle` DataHandle. The TCPSocketHandle ctor copies
    // (ownership transfer) the socket into its own member, so after this call
    // both `socket_` and `tcp_.socket()` are detached (fd = -1) and only the
    // returned DataHandle holds the connection.
    return std::make_unique<DataHandleWrapper>(new eckit::TCPSocketHandle(tcp_.socket()));
}

// ==================== MemoryWriteStreamWrapper ====================

MemoryWriteStreamWrapper::MemoryWriteStreamWrapper() : buf_(4096), mem_(buf_) {}

rust::Slice<const uint8_t> MemoryWriteStreamWrapper::buffer() {
    return {static_cast<const uint8_t*>(buf_.data()), static_cast<size_t>(mem_.bytesWritten())};
}

// ==================== MemoryReadStreamWrapper ====================

MemoryReadStreamWrapper::MemoryReadStreamWrapper(rust::Slice<const uint8_t> data) :
    buf_(data.data(), data.size()), mem_(buf_) {}

// ==================== Factory functions ====================

std::unique_ptr<StreamWrapper> stream_connect(rust::Str host, int32_t port) {
    return std::make_unique<TcpStreamWrapper>(std::string(host), port);
}

std::unique_ptr<StreamWrapper> stream_memory_write() {
    return std::make_unique<MemoryWriteStreamWrapper>();
}

std::unique_ptr<StreamWrapper> stream_memory_read(rust::Slice<const uint8_t> data) {
    return std::make_unique<MemoryReadStreamWrapper>(data);
}

}  // namespace eckit_bridge
